src/wags_tails/utils/storage.py

```diff
@@ -7,42 +7,144 @@
+
+    # since we might be making multiple new subdirectories, use the nearest existing directory
+    probe = data_dir
+    while not probe.exists():
+        parent = probe.parent
+        if parent == probe:
+            break
+        probe = parent
+
+    if not probe.exists() or not probe.is_dir():
+        msg = base_failure_msg.replace(
+            "INSERT_SPECIFIC_HERE",
+            "Parent directory doesn't exist or isn't a directory.",
+        )
+        _logger.error(msg)
+        raise WagsTailsDirWriteError(msg)
+    if not os.access(probe, os.W_OK | os.X_OK):
+        msg = base_failure_msg.replace(
+            "INSERT_SPECIFIC_HERE",
+            "os.access() check indicates user lacks write permissions.",
+        )
+        _logger.error(msg)
+        raise WagsTailsDirWriteError(msg)
+    if hasattr(os, "statvfs"):
+        try:
+            if os.statvfs(probe).f_flag & getattr(os, "ST_RDONLY", 1):
+                msg = base_failure_msg.replace(
+                    "INSERT_SPECIFIC_HERE", "Filesystem appears to be read-only."
+                )
+                _logger.error(msg)
+                raise WagsTailsDirWriteError(msg)
+        except OSError:
+            pass
+    try:
+        data_dir.mkdir(exist_ok=True, parents=True)
+    except PermissionError as e:
+        msg = base_failure_msg.replace("INSERT_SPECIFIC_HERE", f"mkdir() failed: {e}.")
+        _logger.exception(msg)
+        raise WagsTailsDirWriteError(msg) from e
+
+
+def get_data_dir(readonly: bool | None = None) -> Path:
     """Get base wags-tails data storage location.
 
     By default, conform to `XDG Base Directory Specification <https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html>`_,
     unless a directory is specified otherwise:
 
     1) check env var ``"WAGS_TAILS_DIR"``
-    2) check env var ``"XDG_DATA_HOME"``
-    3) check env var ``"XDG_DATA_DIRS"`` for a colon-separated list, skipping any
-        that can't be used (i.e. they're already a file)
-    4) otherwise, use ``~/.local/share/``
+    2) check env var ``"XDG_DATA_HOME"``. If set, use ``${XDG_DATA_HOME}/wags_tails/``
+    3) check env var ``"XDG_DATA_DIRS"`` for a colon-separated list, looking for an
+        element that contains a ``wags_tails/`` subdirectory (only available in read-only mode)
+    4) otherwise, use ``~/.local/share/wags_tails``
+
+    Enable read-only mode by calling with ``readonly=True`` or setting the env var ``WAGS_TAILS_READONLY=TRUE``.
+
+    * If read-only is enabled, ``$XDG_DATA_DIRS`` can be used to provide a data directory, but only if
+      a ``wags-tails`` subdirectory already exists within it. Otherwise, an individual directory entry
+      is skipped. If unable to resolve to a directory that exists, raises ``WagsTailsDirNotAvailableError``.
```

Comment:
Is this extra criteria needed? Only if read-only is enabled, and only if the subdirectory already exists?